crypto/dh/dh_check.c

```diff
@@ -18,7 +18,39 @@
  * p is odd
  * 1 < g < p - 1
  */
+int DH_check_params_ex(const DH *dh)
+{
+    int ok = 0;
+    BIGNUM *tmp = NULL;
+    BN_CTX *ctx = NULL;
+
+    ctx = BN_CTX_new();
+    if (ctx == NULL)
+        goto err;
+    BN_CTX_start(ctx);
+    tmp = BN_CTX_get(ctx);
+    if (tmp == NULL)
+        goto err;
+
+    if (!BN_is_odd(dh->p))
+        DHerr(DH_F_DH_CHECK_PARAMS_EX, DH_R_CHECK_P_NOT_PRIME);
+    if (BN_is_negative(dh->g) || BN_is_zero(dh->g) || BN_is_one(dh->g))
+        DHerr(DH_F_DH_CHECK_PARAMS_EX, DH_R_NOT_SUITABLE_GENERATOR);
+    if (BN_copy(tmp, dh->p) == NULL || !BN_sub_word(tmp, 1))
+        goto err;
+    if (BN_cmp(dh->g, tmp) >= 0)
+        DHerr(DH_F_DH_CHECK_PARAMS_EX, DH_R_NOT_SUITABLE_GENERATOR);
 
+    ok = 1;
+ err:
+    if (ctx != NULL) {
+        BN_CTX_end(ctx);
+        BN_CTX_free(ctx);
+    }
+    return ok;
+}
```

Comment:
Why this code duplication?  I'd rather see it written in terms of `DH_check_params`, perhaps something like this?

``` C
int DH_check_params_ex(const DH *dh)
{
    int errflags = 0;
    int ok = DH_check_params(dh, &errflags);

    if ((errflags & DH_CHECK_P_NOT_PRIME) != 0)
        DHerr(DH_F_DH_CHECK_PARAMS_EX, DH_R_CHECK_P_NOT_PRIME);
    /* ... similar for the other bits */

    return ok;
}
```

Reply:
Ahh, I have considered this. But I am not sure if we are going to remove the non-ex version in next API-changeable version, say 1.2.0, a duplicated version will be easier to remove or not...

Reply:
`DH_check_params` has its uses, I think it's a bit presumptuous to call it obsolete (same goes for the others).

I'm not even sure I think the new `DH_check_params_ex` is the right way to go, as I'm not sure I see the diverse ways a key can be invalid as an error per se, especially in the function that's supposed to check for it!  After all, if the checker returns an error, what does that indicate, that there was something wrong with the call of the function (i.e. the pointer is NULL or points at something that's not even a DH structure), or that the key is invalid according to our criteria (i.e. it *is* a DH structure, it just hasn't been filled in right).

So I'm actually questioning the necessity of `DH_check_params_ex` and the others, and am wondering if it wouldn't be better to adapt the new checking API to expect functions to set diverse validity bits in an int.  But then, that requires a bit of thinking, 'cause for now, there are no common bits between the algorithms...

You see where I'm going with this?

Reply:
Yes. Understood.

Reply:
And I found other algo's key-checking functions act in the same way as what `DH_check_params_ex` does - all errors (either 'function' wrong or 'key' wrong) - for instance, `RSA_check_key_ex` will fill the error stack with something like 'P is not prime', 'N does not equal to p multiplies q'... and also the something like 'value missing' or 'malloc failure'... So it seems that it's consistent to handle all 'algorithm-checking' APIs' errors in this way...

Reply:
(And also ECC does so...)

Reply:
((ahem)) the different algorithms implementation weren't necessarily always consistent in all matters.  That's what the EVP interface attempts to do, bring some level of consistency.

Reply:
If there's no need to make algorithms' check function consistent, then I think there's no need to introduce the `_ex` ones here - just do the logic those `_ex` provide in the EVP check implementation of DH module...

Reply:
> just do the logic those _ex provide in the EVP check implementation of DH module...

If you are sure about this, I can change the code...

Reply:
(I just feel it's a little strange to use 'check' functions of different algos in different ways, which will make the application code complicated to deal with different error handling, so this is the original motivation of the introduction of the `DH_xxx_ex` APIs)